### ISSUE TITLE
Reranking listwise loss occur float and bfloat16 conflict

### ISSUE THREAD

**Issue by LetheHS** [UKPLab/sentence-transformers#3793]
I use PListMLELoss/LambdaLoss/... to train a reranker model based on bge-reranker-v2-m3, then at plist_mle.py line 222
`
logits_matrix[batch_indices, doc_indices] = logits
`
Runtime Error: Index put requires the source and destination dtype match, got Float for the destination and BFloat16 for the source.

This maybe a bug, because I find in the following code
`
logits_matrix = torch.full((batch_size, max_docs), 1e-16, device=self.model.device)
`
the type of 1e-16 is float32.

**Comment by Incheonkirin** [comment]
Thanks for the clear report @LetheHS — I've looked into this and a fix is on the way.

One thing worth flagging: the suggested `torch.full(..., dtype=logits.dtype)` (matching the matrix to the logits dtype) isn't quite sufficient on its own. Once `logits_matrix` becomes bf16, the next step that fills in the labels — `labels_matrix[...] = torch.cat(labels).float()` — hits the same error in reverse (a float32 source into a bf16 destination). On top of that, these losses run `exp`/`cumsum`/`log`/`softmax` without a max-shift, which is numerically fragile in bf16 and overflows in fp16.

So the more robust fix is to compute the loss in float32 — upcasting the logits with `.float()` right after the activation. It's a no-op for the usual fp32 training and only changes behavior under bf16/fp16.

This turns out to affect the whole learning-to-rank loss family rather than just `PListMLELoss`: `LambdaLoss`, `ListNetLoss`, and `ADRMSELoss`, plus `ListMLELoss`/`RankNetLoss` via inheritance — 6 in total. I've reproduced the crash and verified the fix together with a small regression test across all of them locally. Opening a PR shortly.

**Comment by tomaarsen** [comment]
Hello!

Overall, it's recommended to load the model in fp32, and then use `bf16=True` in the `CrossEncoderTrainingArguments`. This means that you'll get the optimizer weights in fp32 (which helps training performance) while the model forward is still done in bf16. The model then outputs fp32 logits for the loss as well. It took me a bit to reproduce this because I always use `bf16=True` or `fp16=True`.

I reviewed #3800 and it looks like a good solution. I'll merge it in a bit. Thank you both!

- Tom Aarsen
